docs/framework/wpf/advanced/hosting-win32-content-in-wpf.md

```diff
@@ -1,357 +1,368 @@
+        {
+            ::MSG m = ConvertMessage(msg);
+
+            // Win32's IsDialogMessage() will handle most of our tabbing, but doesn't know
+            // what to do when it reaches the last tab stop
+            if (m.message == WM_KEYDOWN && m.wParam == VK_TAB) {
+                HWND firstTabStop = GetDlgItem(dialog, IDC_EDIT1);
+                HWND lastTabStop = GetDlgItem(dialog, IDCANCEL);
+                TraversalRequest^ request = nullptr;
+
+                if (GetKeyState(VK_SHIFT) && GetFocus() == firstTabStop) {
+                    // this code should work, but there’s a bug with interop shift-tab in current builds
+                    request = gcnew TraversalRequest(FocusNavigationDirection::Last);
+                }
+                else if (!GetKeyState(VK_SHIFT) && GetFocus() == lastTabStop) {
+                    request = gcnew TraversalRequest(FocusNavigationDirection::Next);
+                }
+
+                if (request != nullptr)
+                    return ((IKeyboardInputSink^) this)->KeyboardInputSite->OnNoMoreTabStops(request);
+
+            }
+
+            // Only call IsDialogMessage for keys it will do something with.
+            if (msg.message == WM_SYSKEYDOWN || msg.message == WM_KEYDOWN) {
+                switch (m.wParam) {
+                    case VK_TAB:
+                    case VK_LEFT:
+                    case VK_UP:
+                    case VK_RIGHT:
+                    case VK_DOWN:
+                    case VK_EXECUTE:
+                    case VK_RETURN:
+                    case VK_ESCAPE:
+                    case VK_CANCEL:
+                        IsDialogMessage(dialog, &m);
+                        // IsDialogMessage should be called ProcessDialogMessage --
+                        // it processes messages without ever really telling you
+                        // if it handled a specific message or not
+                        return true;
+                }
+            }
+
+            return false; // not a key we handled
+        }
+```
+
+This is a lot of code in one piece, so it could use some more detailed explanations. First, the code using [!INCLUDE[TLA#tla_cpp](../../../../includes/tlasharptla-cpp-md.md)] and [!INCLUDE[TLA#tla_cpp](../../../../includes/tlasharptla-cpp-md.md)] macros; you need to be aware that there is already a macro named `TranslateAccelerator`, which is defined in winuser.h:
+
+```cpp
+#define TranslateAccelerator  TranslateAcceleratorW
+```
+
+So make sure to define a `TranslateAccelerator` method and not a `TranslateAcceleratorW` method.
+
+Similarly, there is both the unmanaged winuser.h MSG and the managed `Microsoft::Win32::MSG` struct. You can disambiguate between the two using the [!INCLUDE[TLA#tla_cpp](../../../../includes/tlasharptla-cpp-md.md)] `::` operator.
+
+```cpp
+virtual bool TranslateAccelerator(System::Windows::Interop::MSG% msg,
+    ModifierKeys modifiers) override
+{
+    ::MSG m = ConvertMessage(msg);
+```
```

Comment:
we should probably add this to fix the rest of the weird formatting here in preview.
```suggestion
}
```
```